src/compile.cpp

```diff
@@ -5133,120 +5144,147 @@ void compile(MyThread* t,
-                    e,
-                    reinterpret_cast<const char*>(
-                        invocation->template_()->spec()->body().begin())),
-                e->vtable->NewStringUTF(
-                    e,
-                    reinterpret_cast<const char*>(
-                        cast<GcByteArray>(
-                            t,
-                            singletonObject(t,
-                                            invocation->pool(),
-                                            bootstrapArray->body()[1]))
-                            ->body()
-                            .begin())),
-                e->vtable->NewStringUTF(
-                    e,
-                    reinterpret_cast<const char*>(
-                        method->class_()->name()->body().begin())),
-                e->vtable->NewStringUTF(e,
-                                        reinterpret_cast<const char*>(
-                                            method->name()->body().begin())),
-                e->vtable->NewStringUTF(e,
-                                        reinterpret_cast<const char*>(
-                                            method->spec()->body().begin())),
-                kind);
-
-            uint8_t* bytes = reinterpret_cast<uint8_t*>(
-                e->vtable->GetPrimitiveArrayCritical(e, lambda, 0));
-
-            GcClass* lambdaClass
-                = defineClass(t,
-                              roots(t)->appLoader(),
-                              bytes,
-                              e->vtable->GetArrayLength(e, lambda));
-
-            bc->resolver->addClass(
-                t, lambdaClass, bytes, e->vtable->GetArrayLength(e, lambda));
-
-            e->vtable->ReleasePrimitiveArrayCritical(e, lambda, bytes, 0);
-
-            e->vtable->PopLocalFrame(e, 0);
-
-            THREAD_RUNTIME_ARRAY(
-                t, char, spec, invocation->template_()->spec()->length());
-            memcpy(RUNTIME_ARRAY_BODY(spec),
-                   invocation->template_()->spec()->body().begin(),
-                   invocation->template_()->spec()->length());
-
-            GcMethod* target = resolveMethod(
-                t, lambdaClass, "make", RUNTIME_ARRAY_BODY(spec));
-
-            bool tailCall = isTailCall(t, code, ip, context->method, target);
-            compileDirectInvoke(t, frame, target, tailCall);
+            JNIEnv* e;
+            if (bc->hostVM->vtable->AttachCurrentThread(bc->hostVM, &e, 0)
+                == 0) {
+              e->vtable->PushLocalFrame(e, 256);
+
+              jclass lmfClass = e->vtable->FindClass(
+                  e, "java/lang/invoke/LambdaMetafactory");
+              jmethodID makeLambda = e->vtable->GetStaticMethodID(
+                  e,
+                  lmfClass,
+                  "makeLambda",
+                  "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/"
+                  "String;Ljava/"
+                  "lang/String;Ljava/lang/String;Ljava/lang/String;I)[B");
```

Comment:
It's partly my fault, since there are so many levels of indentation and we're limiting clang-format to 80 columns.  It wouldn't be quite as bad if I broke this into several methods.